Alternative glTF coordinate conversion

greeble-dev opened the pull request:
## Objective

Try a different approach to glTF coordinate conversion. The goal is to satisfy some common use cases now while punting on the more complex and controversial parts. This is an alternative fix for https://github.com/bevyengine/bevy/issues/5670. 

#### Why A Draft?

The PR is currently a first pass intended for design discussion. It's functional, but needs more testing, clean-up, documentation, and release notes.

## Summary

The Bevy glTF loader can convert nodes and meshes from glTF's "+Z forward" to Bevy's "-Z forward". But there's been debate on whether its approach will work for all users (#19686, #20131).

This PR takes a different approach - it drops the node conversion in favour of a whole scene corrective. I think it satisfies several important use cases while keeping the behaviour simple and reliable. But it doesn't cover every case, and could be a regression for some users. (Note that the current conversion was added during 0.17, so from a 0.16 user's perspective there's no regression or changes in behaviour.)

## Background

There's been confusion over how glTF behaves and what users might want from coordinate conversion. This section recaps the basic concepts, glTF's semantics, the current loader behaviour, and some potential user stories.

### Coordinate Systems and Semantics

3D coordinate systems can have semantics assigned to their axes. These semantics are often defined as a forward axis, an up axis, and a [handedness](https://en.wikipedia.org/wiki/Right-hand_rule) - the side axis is implicit in the other choices.

Bevy's standard semantics are "-Z = forward, +Y = up, right handed". The standard is codified by the `forward` and `up` methods of `Transform` and `GlobalTransform`, and by the renderer's interpretation of camera and light transforms. There are debates about the standard and whether users should be able to choose different semantics. This PR does not account for those debates, and assumes that users want to follow the current standard.

Other engines, DCCs and file formats use different semantics. Unlike Bevy, some vary their semantics by object type (e.g. a camera's forward axis may not be the same as a light's). Some only specify an up axis, leaving the forward and side axes unspecified.

Assets might not follow the standard semantics of their file format. Static mesh hierarchies and skeletal animation rigs may even have per-node or per-joint semantics - a character rig could be +Y forward in the scene, while the head joint is +Z forward. One character rig might have both feet +X forward, while another rig might have the left foot +X forward and the right foot -X forward. This creates complexity, but also creates jobs, so no-one can say if it's good or bad.

Users may want asset loaders that convert assets to Bevy's standard semantics. This is not the only option though. The user could edit the assets themselves or run a conversion script in DCC, but that's a pain - particularly for users who rely on asset packs and don't have DCC experience. Another option is to implement an asset transform that does coordinate conversion, but having the options right there in the loader is convenient.

The conversion options of a loader can be contentious - users have different needs, and assets may have ambiguities than can only be resolved by the user. There will never be a simple "it just works" option, although there could be a least worst default that satisfies the biggest group of users.

### Loaders And User Stories

Bevy currently has a glTF loader, and I'm assuming it will get in-repo FBX and USD loaders at some point. These loaders are likely to follow a common pattern:

- The files contain meshes, which correspond to Bevy `Mesh` assets and skinned meshes.
    - Bevy meshes can only have a single material, so what the file format considers a single mesh might be multiple Bevy meshes.
- The files have a node hierarchy, where nodes roughly correspond to Bevy entities with a `Transform`.
    - Nodes can optionally be mesh instances, cameras, lights or skinned mesh joints.
- The loader outputs the assets and a `Scene` with an entity hierarchy that tries to match the file's node hierarchy.
    - Some aspects of nodes (e.g. pivot transforms) can't be represented in Bevy within a single entity.
    - So a 1:1 mapping might not be possible - instead nodes become multiple entities, or some data is lost (e.g. baking down pivot transforms).
- Users can choose to spawn the scene, or they can ignore it and use the assets directly.

For coordinate conversion in the loader, some user stories might be:

- "I want to spawn a scene on an entity with Bevy semantics and have it look right."
    - This is probably the most common case - the user wants to do `SceneRoot(load("my.gltf"))` and have it visually match the entity's `Transform::forward()`, and cameras and lights should do the right thing.
    - The user might not care about the semantics of mesh assets and nodes in the scene - they just want the scene as a whole to look right.
- "I want to spawn a scene, and convert some or all of the nodes to Bevy semantics."
    - The user might have nodes in their scene that they want to animate manually or hook up to other systems that assume Bevy semantics.
    - That becomes easier if the loader can convert the node's forward to match `Transform::forward()`.
    - Conversely, some users may want nodes to stay as they are (particularly skeletal animation rigs).
- "I want to instance a mesh on an entity with Bevy semantics. I'm not using a scene."
    - Maybe the user is doing `Mesh(load("mesh.gltf#Mesh0"))` and wants it to match the entity's forward.
    - Or it's the first stage of an asset pipeline and the remaining stages expect Bevy semantics.
- "I don't want the loader to touch anything."
    - Maybe they've already converted the file, or want to convert it post-load, or don't want to use Bevy semantics at all.
- "I want one of the other conversion stories, but the loader should convert to my chosen semantics rather than Bevy's".
    - Z-up is not a crime.

### glTF Semantics

glTF [scene semantics](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#coordinate-system-and-units) are "+Z = forward, +Y = up, right handed". This is almost the same as Bevy, except that scene forward is +Z instead of Bevy's -Z.

Some glTF assets do not follow the spec's scene semantics. The Kenney asset packs use a mix of +Z and -Z forward. At least [one of the Khronos sample assets](https://github.com/KhronosGroup/glTF-Sample-Assets/tree/main/Models/Duck) uses +X forward. That said, the majority of Kenney assets and almost all Khronos sample assets that I tested do follow the spec.

glTF [camera node](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#view-matrix) and [light node](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_lights_punctual/README.md#adding-light-instances-to-nodes) semantics are different from glTF scene semantics - they're -Z forward, same as Bevy.

The glTF spec doesn't explicitly say if non-camera/light nodes and mesh buffers have semantics. I'm guessing that some users will have nodes and meshes that follow the spec's scene semantics, and might want them converted to Bevy semantics. But as noted in the user stories, it's likely that other users will have different needs.

glTF and Bevy allow a single node/entity to be both a mesh and a camera or a light. This only makes sense if the user intends the mesh to have the same semantics as cameras and lights. I think it's very unlikely that significant numbers of users will want support for this combination - many other DCCs, file formats and engines don't support it at all.

### How The Bevy glTF Loader Works

The loader maps glTF nodes to Bevy entities. It also adds entities for two cases:

1. A single "scene root" entity is added as a parent of the glTF root nodes.
    - Note that this is not the user's entity with the `SceneRoot` component - the scene root entity is a child of that entity.
2. Mesh primitive entities are added as a child of each glTF mesh node.
    - In glTF, a single mesh node can contain multiple primitives.
    - But in Bevy a mesh component can only contain a single primitive, so one entity can't contain multiple primitives.
    - So, for each primitive, Bevy adds a child entity with a mesh component.

A single branch of the resulting scene hierarchy might look like this:

- User entity with `SceneRoot` component.
    - Scene root entity.
        - glTF root node entity.
            - glTF intermediate node entities.
                - glTF mesh node entity (does not contain `Mesh3d` component)
                    - Mesh primitive entities (does contain `Mesh3d` component).

### glTF Loader Changes In 0.17

In Bevy 0.16, the only user story supported by the glTF loader was "no conversion". During the 0.17 cycle, #19633 and some follow up PRs have implemented an option that converts the mesh buffers, nodes, and node animation tracks.

The changes do satisfy some user stories, including the common "convert scene semantics" (mostly) and "convert mesh semantics". But there's some problems:

- The conversion depends on converting both nodes and meshes.
    - Some users might want to convert the scene without converting nodes and/or meshes.
- Light and camera nodes get complicated.
    - glTF camera/light nodes already match Bevy semantics, so they need a counter-conversion (since their parent might have been converted).
    - Animation tracks for lights and cameras are not correctly converted.
        - (Counterpoint: This is fixable at the cost of some complexity)
    - Child nodes of lights and cameras are not correctly converted.
        - (Counterpoint: Also fixable, and probably a niche case?)
    - The conversion can't support a node that's a mesh instance and also a light and/or a camera.
        - (Counterpoint: As mentioned earlier, this is probably a very niche or non-existent use case.)

There was hope that the conversion option could be enabled by default in 0.18 (https://github.com/bevyengine/bevy/issues/19686). But this now looks uncertain (https://github.com/bevyengine/bevy/pull/20131).

## This PR

The big change in this PR is the removal of node conversion. Instead, the scene root entity and mesh primitive entities that Bevy adds are given corrective transforms.

So taking our example branch from the glTF explanation, the pre-PR behaviour is:

- Scene root entity. 
    - glTF root node entity. <-- CONVERTED
        - glTF intermediate node entities. <-- CONVERTED
            - glTF mesh node entity. <-- CONVERTED
                - Mesh primitive entities.

After this PR:

- Scene root entity. <-- CORRECTIVE
    - glTF root node entity. 
        - glTF intermediate node entities. 
            - glTF mesh node entity.
                - Mesh primitive entities. <-- CORRECTIVE

Mesh buffers and skinned mesh bind poses are still converted largely as before.

- Pros:
    - Simpler - there's no need to convert animations, and no need to special case cameras and lights.
    - The common user stories of "convert scene semantics" and "convert mesh assets" are still satisfied.
    - Adds support for the "don't convert nodes" story.
- Cons:
    - Regresses the "*do* convert nodes" story.

The scene and mesh conversions are exposed as separate options - I expect some users might want one without the other. They're both disabled by default.

```rust
pub struct GltfConvertCoordinates {
    pub scene: bool,
    pub meshes: bool,
}
```

The PR also removes the `gltf_convert_coordinates_default` feature and 0.18 warning message, as there's still debate on what these should do. This aligns with https://github.com/bevyengine/bevy/pull/20131, and it's fine if that PR lands first.

### Testing

I've tested various examples and glTFs with each combination of options.

```
cargo run --example animated_mesh
cargo run --example scene_viewer "assets/models/faces/faces.glb"
```

More testing is needed. In particular I want to test the Trenchbroom cases that motivated the original work.

I would have liked to add an example that spawns a separate scene for each combination of options and supports CI screenshot testing. But this is difficult due to #11111.

### What Happens After This PR?

Adding support for node conversion is the obvious next step. But getting it right will be complicated. I don't have any plans to try it.

I would like to add support for arbitrary semantic conversion. It's not that difficult to implement for glTF, and the code will be useful for other file loaders importers and asset transformers.

```rust
struct CoordinateSystem {
    forward: SignedAxis,
    up: SignedAxis,
    right: SignedAxis,
}

impl CoordinateSystem {
    const BEVY: Self = Self::forward_up(SignedAxis::NEGATIVE_Z, SignedAxis::POSITIVE_Y);
    const GLTF_SCENE: Self = Self::forward_up(SignedAxis::POSITIVE_Z, SignedAxis::POSITIVE_Y);
}

pub struct CoordinateConversion {
    pub from: CoordinateSystem,
    pub to: CoordinateSystem,
}

impl CoordinateConversion {
    const GLTF_SCENE_TO_BEVY: Self = Self::from_to(CoordinateSystem::GLTF_SCENE, CoordinateSystem::BEVY);
}
```

### What About The Forward Flag Proposal?

There's a [proposal](https://github.com/bevyengine/bevy/pull/20135) to allow per-transform semantics, aka the "forward flag". This means the axis of `Transform::forward()` and others would depend on a variable in the `Transform`. In theory the forward flag avoids the need for coordinate conversion in the loader, although how it works in practice is still to be worked out.

Even if the forward flag is implemented, I think this PR is worthwhile.

1. Implementing the forward flag would still need changes to the loader. The options and logic will be similar to what's needed for coordinate conversion, so this PR at least lays some groundwork.
2. Users may still want some form of coordinate conversion. They might be working with raw `Mesh` assets and want them converted, or they might prefer to avoid the forward flag for other reasons. So supporting both could be useful, although it would be the most complex choice.

janhohenheim reviewed (approved):
Did not test this in-depth with all scenes, but the explanation and the code make sense to me. This is a well thought out improvement on the situation on `main` and I feel comfortable merging this, provided we untangle the question of whether it should convert by default or not.

I'm in favor of still nudging the user to opt-in so we can gather feedback before changing the default, but that decision is outside the scope of this PR.

greeble-dev commented:
Changed from draft to ready for review. I've done some more testing, added a migration guide for 0.18, and expanded the `GltfConvertCoordinates` docs.

Igor-dvr reviewed (approved):
Viewed the docs and code.

cart commented on crates/bevy_gltf/src/convert_coordinates.rs:
```diff
@@ -48,34 +30,84 @@ impl ConvertCoordinates for [f32; 4] {
+/// coordinate system. This means cameras and lights will match
+/// `Transform::forward` even if conversion is disabled.
+#[derive(Copy, Clone, Default, Debug, Serialize, Deserialize)]
+pub struct GltfConvertCoordinates {
```
It seems to me like the big downside of this new approach is that local transforms within a scene (that correspond to nodes) will now no longer have "correct" forward directions. This relies on nesting the scene underneath a top level entity that is "uncorrected", which we use to call "forward" on. Trying to do that _within_ the scene will not produce the expected "forward" behavior.

This is likely to be problematic in the world of BSN "scene inheritance", where rather than spawning the gltf scene root underneath a higher level "bevy scene root", we would instead treat the gltf scene root as the root, and write our changes directly on top. I believe this is the better model when compared to the current "add another root on top" approach, as it allows developers to think about and interact with GLTF scenes exactly as if they were bevy scenes (ex: if the root node has a property, then when we inherit from it, we can directly read/write that property on the root).

I believe this will also come into play if developers try to animate the root transform of the scene, which will snap our corrections back to whatever the animation thinks is correct.

The benefit of course is that the scene is "locally consistent". But within that space, everything is still "uncorrected". If a "player" node exists inside a larger GLTF "world" scene, that player scene's forward will be wrong.

janhohenheim replied:
Just want to note that I fully agree with what @greeble-dev says here, including the part of wanting to not keep two separate options. I would prefer to yeet the quirky version in favor of this, and I hope that has some weight given that the quirky version is what I upstreamed :)

cart replied:
> The "scene conversion" option in this PR is basically "do whatever is necessary to make it visually correct but without changing node semantics". I think that can work however BSN organizes the hierarchy

The point is that "do whatever" currently relies on adding an additional layer of hierarchy when spawning in Bevy to provide the "correct" forward direction. If we remove that, then we can't "do whatever" to fix it without changing the node semantics. This isn't a problem _currently_, but it will surface the moment we remove that added layer of hierarchy (and I strongly think that we should).

> Just want to note that I fully agree with what @greeble-dev says here

I'm cool with "no node conversion, reliable scene conversion" path for 0.18. I just expect us to very quickly feel the need to reopen this conversation (and reintroduce essentially all of the existing node conversion code) when BSN lands and we drop the extra hierarchy layer.

> Semantics: "scene conversion"

I don't love this framing for the current implementation. I think a more functional description like `rotate_scene_root` would help people conceptualize this better than something like "scene conversion" which could be interpreted to be a more holistic conversion.

cart replied:
Just pushed a proposed `rotate_scene_entity` (this matches the terminology in the docs and solves the "root ambiguity" issue) and `rotate_meshes` rename. I really don't think high level / general purpose names are appropriate here. Open to alternatives though!

I also added some additional clarity to the `rotate_scene_entity` docs.
